### ISSUE TITLE
Running out of memory when running `aclocal` despite having plenty allocated

### ISSUE THREAD

**Issue by dantswain** [docker/for-mac#495]
### Expected behavior

Build script executes succesfully.

### Actual behavior

Build script runs out of memory, often during execution of `aclocal`.

See https://forums.docker.com/t/huge-com-docker-hyperkit-memory-usage-and-aclocal-failure/21101/2 where it was recommended that I create an issue here.

### Information

Diagnostic ID: 3129F357-D201-4F0A-8225-01ABAF1D471A
Docker for Mac: 1.12.0-a (Build 11213)
macOS: Version 10.11.6 (Build 15G31)
[OK]     docker-cli
[OK]     app
[OK]     moby-syslog
[OK]     disk
[OK]     virtualization
[OK]     system
[OK]     menubar
[OK]     osxfs
[OK]     db
[OK]     slirp
[OK]     moby-console
[OK]     logs
[OK]     vmnetd
[OK]     env
[OK]     moby
[OK]     driver.amd64-linux

### Steps to reproduce

It's difficult for me to produce easily reproducible steps because this is part of a build script for proprietary software.  It tends to fail during a compile of [libuv](https://github.com/libuv/libuv) while executing `aclocal`.  During the build, the memory usage of 'com.docker.hyperkit' grows very large - several gigabytes, but always less than what I have configured in docker preferences.

I am able to run this same build script on a linux host running docker and it completes successfully.

**Comment by mchiang0610** [comment]
@dantswain I see that you've tried allocating 14GB of memory previously and it still failed. 

Would it be possible for you to give us the compose or dockerfile used for us to try and reproduce this? 

Thanks!

**Comment by dantswain** [comment]
Hi @mchiang0610 unfortunately there's a lot of proprietary stuff tangled up in there.  I can try to come up with a minimal example, but it will take some time.

I can give a rough outline:
* the "build_image" image is gentoo-based with cmake, git, ccache, and libuv installed
* `docker run --rm -v ${PWD}:/app -v ${LOCAL_CCACHE}:/ccache -w /app/build build_image cmake ..`
* `docker run --rm -v ${PWD}:/app -v ${LOCAL_CCACHE}:/ccache -w /app/build build_image make`

My semi-educated theory is that it might have something to do with the number of files in `${PWD}` and/or `${LOCAL_CCACHE}`.  

```
$ find . -type f | wc -l
   84540
```

53k of those come from boost - it's hard for me to cut that number down much :/

My only other vague theory was that logging might be involved - the build process creates quite a lot of stdout.

Anyway, I'll try to come up with a minimal example that doesn't involve anything proprietary.

**Comment by dsheets** [comment]
@dantswain could you please run

```sh
while true; do date && ps ucx | grep -i docker | awk '{ print $4 "\t" $6 "\t" $11 }' && vm_stat && echo && sleep 1; done
```

while executing your build and report at least the initial baseline reading and some readings just prior to the failure?

**Comment by dantswain** [comment]
@dsheets Done!  I can upload the whole thing if you want. 

Here is the first entry, before I started:
```
Fri Sep  2 09:52:09 EDT 2016
0.3	52512	Docker
0.0	3888	com.docker.osx.hyperkit.linux
0.0	7712	com.docker.osx.hyperkit.linux
0.2	26896	com.docker.db
0.0	408	com.docker.osx.hyperkit.linux
0.0	2764	com.docker.osx.hyperkit.linux
0.0	5248	com.docker.frontend
0.0	5508	com.docker.frontend
0.8	127564	com.docker.hyperkit
0.1	9052	com.docker.driver.amd64-linux
0.1	14604	com.docker.driver.amd64-linux
0.1	9676	com.docker.slirp
0.1	10132	com.docker.osxfs
Mach Virtual Memory Statistics: (page size of 4096 bytes)
Pages free:                              644857.
Pages active:                           1559213.
Pages inactive:                          996208.
Pages speculative:                        15740.
Pages throttled:                              0.
Pages wired down:                        673083.
Pages purgeable:                          46813.
"Translation faults":                2356762083.
Pages copy-on-write:                   85260403.
Pages zero filled:                   1083045564.
Pages reactivated:                   5157012494.
Pages purged:                          10117079.
File-backed pages:                       854776.
Anonymous pages:                        1716385.
Pages stored in compressor:             1518635.
Pages occupied by compressor:            303457.
Decompressions:                       398045944.
Compressions:                         495463107.
Pageins:                               22560626.
Pageouts:                               1362686.
Swapins:                              293160006.
Swapouts:                             301638126.
```

Here is the final (I'm not sure exactly when the job failed... this might have been a few minutes afterwards when I noticed my CPU fan slow down and checked on my build):

```
Fri Sep  2 10:59:42 EDT 2016
17.1	2875852	com.docker.hyperkit
0.2	29576	Docker
0.0	5364	com.docker.driver.amd64-linux
0.1	8612	com.docker.driver.amd64-linux
0.0	2932	com.docker.slirp
0.0	4216	com.docker.osxfs
0.0	3816	com.docker.osx.hyperkit.linux
0.0	6100	com.docker.osx.hyperkit.linux
0.0	4324	com.docker.db
0.0	252	com.docker.osx.hyperkit.linux
0.0	1244	com.docker.osx.hyperkit.linux
0.0	5108	com.docker.frontend
0.0	5540	com.docker.frontend
Mach Virtual Memory Statistics: (page size of 4096 bytes)
Pages free:                               25846.
Pages active:                           1605375.
Pages inactive:                         1605028.
Pages speculative:                          125.
Pages throttled:                              0.
Pages wired down:                        708274.
Pages purgeable:                          57762.
"Translation faults":                2410512447.
Pages copy-on-write:                   88150472.
Pages zero filled:                   1108779017.
Pages reactivated:                   5295580322.
Pages purged:                          10387662.
File-backed pages:                       923008.
Anonymous pages:                        2287520.
Pages stored in compressor:             4856304.
Pages occupied by compressor:            247700.
Decompressions:                       407169594.
Compressions:                         509697566.
Pageins:                               23155097.
Pageouts:                               1473903.
Swapins:                              296248818.
Swapouts:                             305457243.
```

I can upload the whole file somewhere if that would help.  It's 5.8M uncompressed.

**Comment by dantswain** [comment]
OK!  I found a simple reproducible way to trigger this situation that doesn't involve any proprietary stuff:

```
mkdir attest
cd attest
docker pull gentoo/stage3-amd64
git clone https://github.com/libuv/libuv
docker run --rm -v ${PWD}:/work -w /work/libuv gentoo/stage3-amd64 sh autogen.sh
docker run --rm -v ${PWD}:/work -w /work/libuv gentoo/stage3-amd64 sh autogen.sh
```

It seems to explode on the second run of `sh autogen.sh`.  I am able to run this without problems on a linux host running `Docker version 1.12.0-rc2, build 906eacd`.  A colleague with a Mac running `Docker version 1.11.0, build 4dc5990` was also able to run this without problems.  Another colleague with a Mac running `Docker version 1.12.1, build 23cf638, experimental` gets the same failure that I do with "Out of memory!".

**Comment by eliwjones** [comment]
I am seeing this as well.

We use a Docker container to do Yocto Bitbake builds (docker run -it bash and do bitbaking).

It works fine when I run the docker image on my Ubuntu laptop and do builds.

But, when my colleagues attempt to run builds using the current Docker for Mac release, they get the "Out of memory!" error while certain recipes are building (say, during an autoreconf). 

And, with previous Beta releases of Docker for Mac, this never happened.

**Comment by jeremyong** [comment]
+1 Also seeing this with builds that use libtoolize or autoconf

**Comment by dantswain** [comment]
@eliwjones @jeremyong What images are you using for builds?  I'm wondering if it's specific to the gentoo image or not..

**Comment by eliwjones** [comment]
@dantswain: I am using `FROM ubuntu:12.04`

**Comment by jeremyong** [comment]
@dantswain debian:stretch for me.

**Comment by itsgrimetime** [comment]
@dantswain I'm using a custom AmazonLinux (RHEL 5) image and am hitting the same issue with a Maven build that uses libtoolize

**Comment by dsheets** [comment]
I've reproduced this issue and then fixed it upstream. After some QA, it should reach Beta 28 in a week or two.

**Comment by dsheets** [comment]
I should also mention, thanks @dantswain for the great work developing a reproduction and thanks to everyone for reporting the issue and helping out. The bug turned out to be related to specific inotify events being delivered on a file that was also being read which ended up causing us to corrupt the Linux node table. Without your work, this bug would have lurked for much longer. Thank you! ❤️

**Comment by dsheets** [comment]
This should now be fixed in the just-released Beta 28. Please give it a try and feel free to re-open this issue or comment here if the problem has not been resolved. The fix should reach the stable channel with 1.12.2 in a week or two. If you experience other problems, please open a new issue. Thanks for using Docker for Mac!

**Comment by dantswain** [comment]
I will test as soon as I can!
